File: spring-boot-oauth2-auth-server/src/main/java/com/bittokazi/oauth2/auth/server/app/services/CustomUserDetailsService.java
package com.bittokazi.oauth2.auth.server.app.services;

import java.util.Optional;
import java.util.stream.Collectors;


import com.bittokazi.oauth2.auth.server.app.models.tenant.User;
import com.bittokazi.oauth2.auth.server.app.models.tenant.security.RoleOauth;
import com.bittokazi.oauth2.auth.server.app.models.tenant.security.UserOauth;
import com.bittokazi.oauth2.auth.server.app.repositories.tenant.OauthClientRepository;
import com.bittokazi.oauth2.auth.server.app.repositories.tenant.UserRepository;
import com.bittokazi.oauth2.auth.server.database.MultiTenantConnectionProviderImpl;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;

/**
 * @author Bitto Kazi
 */

@Transactional
public class CustomUserDetailsService implements UserDetailsService {

    private UserRepository userRepository;

    private OauthClientRepository oauthClientRepository;

    private MultiTenantConnectionProviderImpl multiTenantConnectionProviderImpl;

    private RegisteredClientRepository registeredClientRepository;

    public CustomUserDetailsService(UserRepository userRepository, OauthClientRepository oauthClientRepository,
                                    MultiTenantConnectionProviderImpl multiTenantConnectionProviderImpl,
                                    RegisteredClientRepository registeredClientRepository) {
        this.userRepository = userRepository;
        this.oauthClientRepository = oauthClientRepository;
        this.multiTenantConnectionProviderImpl = multiTenantConnectionProviderImpl;
        this.registeredClientRepository = registeredClientRepository;
    }

    private static final Logger logger = LoggerFactory.getLogger(CustomUserDetailsService.class);

    @Override
    public UserDetails loadUserByUsername(String username) {
        try {
            Optional<User> user = userRepository.findOneByUsernameIgnoreCase(username);
            if (!user.isPresent()) {
                throw new UsernameNotFoundException("user not found");
            }
            UserOauth userDetails = new UserOauth();
            userDetails.setUsername(user.get().getUsername());
            userDetails.setPassword(user.get().getPassword());
            userDetails.setRoles(user.get().getRoles().stream().map(role -> {
                RoleOauth roleOauth = new RoleOauth(role.getId(), role.getName(), role.getTitle());
                return roleOauth;
            }).collect(Collectors.toSet()));
            userDetails.setEmail(user.get().getEmail());
            userDetails.setAuthorities(user.get().getAuthorities());
            return userDetails;
        } catch (Exception e) {
            logger.error("ERROR During loading User ", e);
        }
        return null;
    }

}

File: spring-boot-oauth2-auth-server/src/main/kotlin/com/bittokazi/oauth2/auth/server/app/services/CustomUserDetailsService.kt
package com.bittokazi.oauth2.auth.server.app.services

import com.bittokazi.oauth2.auth.server.app.models.tenant.Role
import com.bittokazi.oauth2.auth.server.app.models.tenant.security.RoleOauth
import com.bittokazi.oauth2.auth.server.app.models.tenant.security.UserOauth
import com.bittokazi.oauth2.auth.server.app.repositories.tenant.UserRepository
import jakarta.transaction.Transactional
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component
import java.util.stream.Collectors

/**
 * @author Bitto Kazi
 */
@Transactional
@Component
open class CustomUserDetailsService(
    private val userRepository: UserRepository
) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        try {
            val user = userRepository.findOneByUsernameIgnoreCase(username)
            if (!user.isPresent) {
                throw UsernameNotFoundException("user not found")
            }
            val userDetails = UserOauth()
            userDetails.username = user.get().username
            userDetails.password = user.get().password
            userDetails.roles = user.get().roles.stream().map<RoleOauth> { role: Role ->
                val roleOauth = RoleOauth(role.id, role.name, role.title)
                roleOauth
            }.collect(Collectors.toSet())
            userDetails.email = user.get().email
            userDetails.authorities = user.get().getAuthorities()
            return userDetails
        } catch (e: Exception) {
            logger.error("ERROR During loading User ", e)
        }
        throw UsernameNotFoundException("User Not Found")
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(CustomUserDetailsService::class.java)
    }
}
